modules/container/container.go

```diff
@@ -0,0 +1,219 @@
+	if err != nil {
+		return id, err
+	}
+	defer client.Close()
+
+	args, _ := shlex.Split(entrypoint)
+
+	root, err := c.flister.Mount(flist, "")
+	if err != nil {
+		return id, err
+	}
+
+	defer func() {
+		if err != nil {
+			c.flister.Umount(root)
+		}
+	}()
+
+	opts := []oci.SpecOpts{
+		oci.WithDefaultSpecForPlatform("linux/amd64"),
+		oci.WithRootFSPath(root),
+		oci.WithProcessArgs(args...),
+		oci.WithEnv(env),
+
+		// NOTE: the hooks run inside runc namespace
+		// it means that we can't do the unmount of the
+		// root fs from here.
+
+		// withHooks(specs.Hooks{
+		// 	Poststop: []specs.Hook{
+		// 		{
+		// 			Path: "umount",
+		// 			Args: []string{path},
+		// 		},
+		// 	},
+		// }),
+	}
+
+	if len(network.Namespace) != 0 {
+		opts = append(
+			opts,
+			getNetworkSpec(network),
+		)
+	}
+
+	for _, mount := range mounts {
+		opts = append(
+			opts,
+			oci.WithMounts([]specs.Mount{
+				{
+					Destination: mount.Target,
+					Type:        mount.Type,
+					Source:      mount.Source,
+					Options:     mount.Options,
+				},
+			}),
+		)
+	}
+
+	ctx := namespaces.WithNamespace(context.Background(), ns)
+
+	container, err := client.NewContainer(
+		ctx,
+		name,
+		containerd.WithNewSpec(opts...),
+	)
+
+	if err != nil {
+		return id, err
+	}
+
+	defer func() {
+		if err != nil {
+			container.Delete(ctx, containerd.WithSnapshotCleanup)
+		}
+	}()
+
+	logs := path.Join("/var/log", ns)
+	os.MkdirAll(logs, 0755)
+	task, err := container.NewTask(ctx, cio.LogFile(path.Join(logs, fmt.Sprintf("%s.log", container.ID()))))
+	if err != nil {
+		return id, err
+	}
+
+	// call start on the task to execute the redis server
+	if err := task.Start(ctx); err != nil {
+		return id, err
+	}
+
+	return modules.ContainerID(container.ID()), nil
+}
+
+func (c *containerModule) Inspect(ns string, id modules.ContainerID) (modules.ContainerInfo, error) {
+	return modules.ContainerInfo{}, nil
+}
+
+func (c *containerModule) Delete(ns string, id modules.ContainerID) error {
+	client, err := containerd.New(c.containerd)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	ctx := namespaces.WithNamespace(context.Background(), ns)
+
+	container, err := client.LoadContainer(ctx, string(id))
+	if err != nil {
+		return err
+	}
+
+	task, err := container.Task(ctx, nil)
+	if err == nil {
+		// err == nil, there is a task running inside the container
+		exitC, err := task.Wait(ctx)
+		if err != nil {
+			return err
+		}
+		trials := 3
+	loop:
+		for {
+			signal := syscall.SIGTERM
+			if trials <= 0 {
+				signal = syscall.SIGKILL
+			}
+			task.Kill(ctx, signal)
+			trials--
+			select {
+			case <-exitC:
+				break loop
+			case <-time.After(1 * time.Second):
+			}
+		}
+
+		if _, err := task.Delete(ctx); err != nil {
+			return err
+		}
+	}
+	spec, err := container.Spec(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := container.Delete(ctx); err != nil {
+		return err
+	}
+
+	return syscall.Unmount(spec.Root.Path, syscall.MNT_DETACH)
```

Comment:
```suggestion
	return c.flister.Umount(spec.Root.Path)
```

Reply:
why u wanna drop the `DETACH` it will make the unmount. this will allow the system to actually perform the unmount once the files are not used anymore. instead of getting an  (filesystem busy) error

Reply:
Look again, I'm not doing a `syscall.Unmount`, but call the Umount method of the flister module.